### DISCUSSION TITLE
Is it legal to call functions from libSystem.Native?

### DISCUSSION THREAD

**Discussion by sakno** [dotnet/runtime#71408] [Q&A] [answered]
.NET 6 provides awesome way to call some PAL functions via PInvoke even in case of single-file deployment. For instance, I have the following code:
```csharp
[DllImport("libSystem.Native", EntryPoint = "SystemNative_GetPwUidR", SetLastError = false)]
internal static extern int GetPwUidR(uint uid, out Passwd pwd, ref byte buf, int bufLen); // buf must be pinned, I know
```

The immediate caller of this function is guarded by `OperatingSystem.IsLinux()` condition. According to #41299 and #44505, CLR interprets such an import as a statically linked function. So my questions are:
* Is this legal for use in libraries outside of .NET repo?
* Is this really compatible with any deployment mode, including AOT?

**Comment by stephentoub** [comment, marked as answer]
> Is this legal for use in libraries outside of .NET repo?

libSystem.Native is implementation detail. It is not supported for use outside of the shared framework itself.

**Reply by sakno** [reply]
I understand that. Anyway, if I accept the risk of incompatible changes in the future, is it working properly with AOT?

**Reply by gfoidl** [reply]
You're on thin ice anyway, so why don't just try it?
And "properly" in the context of unsupported scenarios is the wrong term 😉.

**Comment by Xyncgas** [comment]
It depends on the states, also it might not work 😂
